core/shared/src/main/scala/zio/Semaphore.scala

```diff
@@ -117,41 +117,59 @@ object Semaphore {
               }
             }
 
-        def restore(promise: Promise[Nothing, Unit], n: Long)(implicit trace: Trace): UIO[Any] =
-          ref.modify {
-            case Left(queue) =>
-              queue
-                .find(_._1 == promise)
-                .fold(releaseN(n) -> Left(queue)) { case (_, permits) =>
-                  releaseN(n - permits) -> Left(queue.filter(_._1 != promise))
+        def restore(promise: Promise[Nothing, Unit], n: Long)(implicit trace: Trace): UIO[Any] = ZIO.suspendSucceed {
+          if (promise.unsafe.isDone) {
+            // If the promise completed, then there isn't any need to remove it from the queue,
+            // since it was removed before it was completed in another release.
+            releaseN(n)
+          } else {
+            // If isDone is false, that means the fiber waiting for the promise was interrupted.
+            // We should therefore remove the promise from the queue, so that a later release doesn't
+            // true to fulfill the promise when nothing is waiting for it.
+            ref.modify {
```

Comment:
I think at this point, perhaps a `Set` or `Map` should be utilized to store the waiters.

Then you can filter out by (a) first checking `.contains`, and then (b) actually doing the filtering if necessary.

Reply:
Then we would lose the order that permits were requested in. I'm not sure  if that is acceptable or not.

The documentation doesn't currently say anything about fairness, but there could be code that relies on that.